x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/SamlServiceProviderDocument.java

```diff
@@ -0,0 +1,404 @@
+            parser -> parser.currentToken() == XContentParser.Token.VALUE_NULL ? null : parser.longValue(),
+            Fields.AUTHN_EXPIRY, ObjectParser.ValueType.LONG_OR_NULL);
+        DOC_PARSER.declareStringArray(SamlServiceProviderDocument::setSigningCertificates, Fields.SIGNING_CERT);
+
+        DOC_PARSER.declareObject(NULL_CONSUMER, (parser, doc) -> PRIVILEGES_PARSER.parse(parser, doc.privileges, null), Fields.PRIVILEGES);
+        PRIVILEGES_PARSER.declareStringOrNull(Privileges::setApplication, Fields.Privileges.APPLICATION);
+        PRIVILEGES_PARSER.declareString(Privileges::setResource, Fields.Privileges.RESOURCE);
+        PRIVILEGES_PARSER.declareStringOrNull(Privileges::setLoginAction, Fields.Privileges.LOGIN_ACTION);
+        PRIVILEGES_PARSER.declareField(Privileges::setGroupActions,
+            (parser, ignore) -> parser.currentToken() == XContentParser.Token.VALUE_NULL ? null : parser.mapStrings(),
+            Fields.Privileges.GROUPS, ObjectParser.ValueType.OBJECT_OR_NULL);
+
+        DOC_PARSER.declareObject(NULL_CONSUMER, (p, doc) -> ATTRIBUTES_PARSER.parse(p, doc.attributeNames, null), Fields.ATTRIBUTES);
+        ATTRIBUTES_PARSER.declareString(AttributeNames::setPrincipal, Fields.Attributes.PRINCIPAL);
+        ATTRIBUTES_PARSER.declareStringOrNull(AttributeNames::setEmail, Fields.Attributes.EMAIL);
+        ATTRIBUTES_PARSER.declareStringOrNull(AttributeNames::setName, Fields.Attributes.NAME);
+        ATTRIBUTES_PARSER.declareStringOrNull(AttributeNames::setGroups, Fields.Attributes.GROUPS);
+    }
+
+    public static SamlServiceProviderDocument fromXContent(String docId, XContentParser parser) throws IOException {
+        SamlServiceProviderDocument doc = new SamlServiceProviderDocument();
+        doc.setDocId(docId);
+        return DOC_PARSER.parse(parser, doc, doc);
+    }
+
+    public ValidationException validate() {
+        final ValidationException validation = new ValidationException();
+        if (Strings.isNullOrEmpty(name)) {
+            validation.addValidationError("field [" + Fields.NAME.getPreferredName() + "] is required, but was [" + name + "]");
+        }
+        if (Strings.isNullOrEmpty(entityId)) {
+            validation.addValidationError("field [" + Fields.ENTITY_ID.getPreferredName() + "] is required, but was [" + entityId + "]");
+        }
+        if (Strings.isNullOrEmpty(acs)) {
+            validation.addValidationError("field [" + Fields.ACS.getPreferredName() + "] is required, but was [" + acs + "]");
+        }
+        if (created == null) {
+            validation.addValidationError("field [" + Fields.CREATED_DATE.getPreferredName() + "] is required, but was [" + created + "]");
+        }
+        if (lastModified == null) {
+            validation.addValidationError(
+                "field [" + Fields.LAST_MODIFIED.getPreferredName() + "] is required, but was [" + lastModified + "]");
```

Comment:
Is there a reason you prefer that change?
I like the existing code because the error message is isolated from potential changes in the `if` condition.
e.g. If we decide that we need to check `lastModified.toEpochMilli() != 0` then the error message would be misleading (because "required" isn't a great descriptor, but it would at least print out the correct current value.

The suggestion saves the tiny cost of a string addition, but risks printing incorrect information if it gets out of sync with the if clause.

Reply:
I'm fine with it, it's just how I would have done it - no need to iterate on this ;)